designs/2018-processors-improvements/README.md

```diff
@@ -0,0 +1,271 @@
+
+End-users can also mix and match processors, such as in this example:
+
+```js
+module.exports = {
+    plugins: ["markdown", "vue", "html"],
+    parserOptions: {
+        ecmaVersion: 2018
+    },
+    overrides: [
+        {
+            files: ["*.md"],
+            processors: ["markdown/markdown", "vue/vue"]
+        },
+        {
+            files: ["*.htm", "*.html"],
+            processors: ["html/html", "vue/vue"]
+        }
+    ]
+};
+```
+
+This configuration uses the same `vue` named processor from `eslint-plugin-vue` for both HTML and Markdown files, but only after an appropriate named processor is used to extract the Vue code.
+
+**Note:** It would be update to the Vue processor to understand the difference between different types of code it may be passed.
+
+### Virtual Filenames for Code Blocks
+
+The `preprocess()` method inside of a processor can now return an array containing either strings (the current behavior) or an object with two properties:
+
+* `filePath` - (string) a virtual file path to map the code block to
+* `text` - (string) the source code text
+
+For example:
+
+```js
+module.exports = {
+    preprocess(text, filename) {
+        
+        const {vueText, jsText } = doSomething(text, filename);
+        
+        return [
+            {
+                filePath: filename + ".vue",
+                text: vueText
+            },
+            {
+                filePath: filename + ".js",
+                text: jsText
+            }
+        ]
+    }
+};
+```
+
+This processor returns two code blocks: one that contains Vue code and is given a virtual filename ending with `.vue` and one that contains regular JavaScript code and is given a virtual filename ending with `.js`.
+
+When a `preprocess()` method returns an object with a `filePath` property, `CLIEngine` will call `getConfigForFile()` on the `filePath` property to determine the correct configuration for the code block.
+
+When a `preprocess()` method returns only a string, `CLIEngine` will interpret that as a JavaScript file and call `getConfigForFile()` using the parent file's filename to determine the correction configuration for the code block.
+
+## Changes to `CLIEngine` and `Linter`
+
+Currently, application of processors happens inside of the `Linter#verify()` and `Linter#verifyAndFix()` methods. In order to implement this proposal, application of processors must happen outside of `Linter` because configuration calculation happens outside of `Linter`.
+
+The application of processors must move into `CLIEngine#executeOnText()` and `CLIEngine#executeOnFiles()` to work alongside with the current configuration calculation. That also means `CLIEngine` becomes responsible for recombining and postprocessing results from multiple `Linter#verify()` calls.
```

Comment:
I'm worried about online demos. If the processors logic is removed from `Linter`, the plugins which have processors cannot make their online demos. For example, https://mysticatea.github.io/vue-eslint-demo/ is using processors API.

Reply:
I hope to make a restriction that `preprocess` cannot change the directory that the file belongs instead. As a result, it doesn't need re-looking up configuration files by every preprocess, then it will avoid performance impact.

Reply:
> If the processors logic is moved to CLIEngine, the plugins which have processors cannot use new processors API because of their online demos.

I'm not advocating moving the logic from `Linter` to `CLIEngine`. I'm proposing that we leave the `Linter` logic in place but have something in `CLIEngine` that is used with the CLI. So the CLI would never use the `Linter` logic but online demos would still be able to work the same as today. Does that make sense?

> I hope to make a restriction that preprocess cannot change the directory that the file belongs instead.

That's a good point! Should we make it so `preprocess` can only return a `filename` that disallows the `\` character?

Reply:
> Should we make it so preprocess can only return a filename that disallows the \ character?

Sounds good to me.

> I'm not advocating moving the logic from Linter to CLIEngine. I'm proposing that we leave the Linter logic in place but have something in CLIEngine that is used with the CLI. So the CLI would never use the Linter logic but online demos would still be able to work the same as today. Does that make sense?

Indeed, it will not break the existing online demos. But I think that it means we cannot use the new feature in online demos (it can be a blocker for migration). I had imaged that we can separate the resolving `overrides` logic from the finding config files logic then can move the former into `Linter`.

Current behavior:

1. `CLIEngine` finds the proper config and resolves `overrides`. (depends on `fs`)
1. `CLIEngine` picks up the proper processor from the config.
1. `Linter` applies the processor to code.
1. `Linter` verifies the processed code with the config.

New behavior that I'm imaging:

1. `CLIEngine` finds the proper config. (depends on `fs`)
1. `Linter` applies processors recursively with the config which contain `overrides` property.
1. `Linter` verifies the processed code with the config.

Reply:
OK. I got it that moving of the logic can be a breaking change.
I'm still thinking that the resolving of `overrides` should not depend on `fs`, but I can separate it from this proposal.

Reply:
Sounds good. Thanks @mysticatea.